src/main/java/com/autosale/shop/service/impl/JwtTokenServiceImpl.java

```diff
@@ -0,0 +1,103 @@
+public class JwtTokenServiceImpl implements JwtTokenService {
+
+    private final String SECRET_KEY;
+
+    private final UserService userService;
+
+    private final int ACCESS_TOKEN_LIFETIME;
+
+    private final int REFRESH_TOKEN_LIFETIME;
+
+    public JwtTokenServiceImpl(@Value("${jwt.secret}") String secret_key, UserService userService, @Value("${jwt.access_token_lifetime}") int access_token_lifetime, @Value("${jwt.refresh_token_lifetime}") int refresh_token_lifetime) {
+        SECRET_KEY = secret_key;
+        this.userService = userService;
+        ACCESS_TOKEN_LIFETIME = access_token_lifetime;
+        REFRESH_TOKEN_LIFETIME = refresh_token_lifetime;
+    }
+
+    @Override
+    public JwtTokensDAO generateTokensPair(User user) {
+        if (user.getId() == null) {
+            user = userService.getVerifiedUser(user.getUserName(), user.getPassword());
+        }
+        JwtBuilder jwtBuilder = Jwts.builder().claim("id", user.getId().toString())
+                .claim("username", user.getUserName())
+                .claim("role", user.getRole())
+                .setId(UUID.randomUUID().toString())
+                .setSubject(String.format("%s,%s", user.getId(), user.getUserName()))
+                .setIssuedAt(new Date()).setExpiration(Date.from(Instant.now().plus(ACCESS_TOKEN_LIFETIME, ChronoUnit.MINUTES)))
+                .signWith(Keys.hmacShaKeyFor(SECRET_KEY.getBytes()));
+        String accessToken = jwtBuilder.compact();
+        String refreshToken = jwtBuilder
+                .claim("type", "refresh")
+                .setIssuedAt(new Date()).setExpiration(Date.from(Instant.now().plus(REFRESH_TOKEN_LIFETIME, ChronoUnit.HOURS)))
+                .compact();
+        return new JwtTokensDAO(accessToken,refreshToken);
+    }
+
+
+    @Override
+    public boolean isValidToken(String token) {
+        Claims claims = Jwts.parserBuilder().setSigningKey(SECRET_KEY.getBytes()).build().parseClaimsJws(token).getBody();
+        User user = userService.findById(Integer.parseInt(claims.get("id", String.class)));
+
+        if (user == null) {
+            return false;
+        }
+
+        return claims.get("username", String.class).equals(user.getUserName()) && claims.get("role", String.class).equals(user.getRole().name()) && claims.getExpiration().after(new Date());
+    }
+
+    @Override
+    public Authentication getAuthentication(String token)
+    {
+        Claims claims = Jwts.parserBuilder().setSigningKey(SECRET_KEY.getBytes()).build().parseClaimsJws(token).getBody();
+
+        return new UsernamePasswordAuthenticationToken(claims.get("id", String.class),"",getAuthorities(claims));
+    }
+
+    @Override
+    public JwtTokensDAO generateTokensFromRefreshToken(String refreshToken) throws UnsupportedJwtException {
+        Claims claims = Jwts.parserBuilder().setSigningKey(SECRET_KEY.getBytes()).build().parseClaimsJws(refreshToken).getBody();
+        if (!claims.get("type", String.class).equals("refresh")) {
```

Comment:
it's better to do it this way:
"refresh".equals(......) 
Because it's null safety